### DISCUSSION TITLE
Configure retention time

### DISCUSSION THREAD

**Discussion by ThePoipoi** [strimzi/strimzi-kafka-operator#4336] [Q&A] [answered]
Hello all.
Just a quick question: I'm trying to configure my Kafka cluster to have a custom retention time.
I have found some information about it in the 0.5.0 documentation, but nothing since.

So I have tried to used default parameters from Kafka, as shown below:

```
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
    version: 2.6.0
    replicas: 1
    [...]
    config:
      [...]
      log.retention.ms: 120000
      log.retention.hours: 0
      log.retention.bytes: 10737
      log.delete.retention.ms: 120000
      log.cleaner.delete.retention.ms: 120000
      log.retention.check.interval.ms: 10000
[...]
```

The goal is to remove each remaining topic message after a specified amount of time (in this case, I'm trying to keep messages only for 2 minutes).

Correct me if I'm wrong, but I have understand that the retention period is used only to marks data segments for deletion. And if I want thoses data delete, I have to use the "cleaner".
Am I right ? What do I do wrong (because nothing is never removed) ?

Thanks

**Comment by scholzj** [comment, marked as answer]
Kafka topics have two policies: delete and cleanup. With the cleanup policy, it is keeping the last message for given message key. With the delete policy, it will delete them based on time. But it will always only delete the whole log segments once all messages in them are over the retention time. You can have some control over how often that happens for example by configuring how often new segments are created (using `log.roll.ms`) so that the old segments are closed and can be deleted even when they are not full. But to do this so often might have performance impact.

The configuration you have above should work in general. But:
* Topics might still override these settings. They are basically just the defaults when not set in the topic.
* The `log.retention.ms` and `log.retention.hours` conflict with each other. Use just one of them.

But the settings you have there would not do what you want. If you want to have some strict time-to-live / message-expiration style behaviour, I do not think Kafka is the right solution.

**Reply by ThePoipoi** [reply]
Thanks for your answer.
No, I do not need a real and precise way to remove messages from Kafka. 
But we don't want to have too much data in the time, so we have to clean up sometimes, after approximately 1h for example.

And it's where I don't succeed !
If I understand correctly, my configuration should done this if there is no other config in the Topic side.
To be honest, we didn't work on the Topic side, and each topic is currently created dynamically by our code.

Here is the log of my last test, with the below configuration :

```
2021-01-29 11:33:09,333 INFO Created log for partition syncTopicSDE-1 in /var/lib/kafka/data/kafka-log0/syncTopicSDE-1 with properties {compression.type -> producer, min.insync.replicas -> 1, message.downconversion.enable -> true, segment.jitter.ms -> 0, cleanup.policy -> [delete], flush.ms -> 9223372036854775807, retention.ms -> 120000, segment.bytes -> 1073741824, flush.messages -> 9223372036854775807, message.format.version -> 2.6-IV0, max.compaction.lag.ms -> 9223372036854775807, file.delete.delay.ms -> 60000, max.message.bytes -> 1048588, min.compaction.lag.ms -> 0, message.timestamp.type -> CreateTime, preallocate -> false, index.interval.bytes -> 4096, min.cleanable.dirty.ratio -> 0.5, unclean.leader.election.enable -> false, retention.bytes -> 10737, delete.retention.ms -> 120000, segment.ms -> 604800000, message.timestamp.difference.max.ms -> 9223372036854775807, segment.index.bytes -> 10485760}. (kafka.log.LogManager) [data-plane-kafka-request-handler-7]
2021-01-29 11:33:09,334 INFO [Partition syncTopicSDE-1 broker=0] No checkpointed highwatermark is found for partition syncTopicSDE-1 (kafka.cluster.Partition) [data-plane-kafka-request-handler-7]
2021-01-29 11:33:09,335 INFO [Partition syncTopicSDE-1 broker=0] Log loaded for partition syncTopicSDE-1 with initial high watermark 0 (kafka.cluster.Partition) [data-plane-kafka-request-handler-7]
2021-01-29 11:33:09,335 INFO [Broker id=0] Leader syncTopicSDE-1 starts at leader epoch 0 from offset 0 with high watermark 0 ISR [0] addingReplicas [] removingReplicas []. Previous leader epoch was -1. (state.change.logger) [data-plane-kafka-request-handler-7]

```

As we can see:
- retention.ms is corretly set
- same for retention.bytes, but not for segment.bytes that (I assume) use the default value
- same for delete.retention.ms
- cleanup.policy is set to 'delete'
- segment.ms use the default '7d' value, and I don't know how to change that from the Kafka configuration
- [...]

So can you tell me what am I doing wrong, and why my messages are still remaining in our Topics ?
Because even if it's not precise, we need to have a cleanup system, and I'm really don't understand what I'm missing.

Thanks Jakub for your time :)

**Reply by scholzj** [reply]
As I said before - `segment.ms` corresponds to `log.roll.ms` in the broker config. If a new segment is rolled out only every 7 days, you would need to send a lot of messages to fill the segment quickly and have them deleted. Also, the delete retention is relevant for the clean-up policy. So it will not help you.

**Reply by ThePoipoi** [reply]
> As I said before - `segment.ms` corresponds to `log.roll.ms` in the broker config.

Ok, I wasn't sure which Kafka parameter correspond to Topic parameter.
And you're totally right, now I also have a specified value for the "segment.ms" property.
But it's still remove nothing from my Topic...So, what I need to do more ?

Thanks for your help.

**Reply by scholzj** [reply]
So can you share how your config looks like right now? Also, if you exec into one of the brokers, you can check in `/var/lib/kafka` directory ... do you see the new segments being created?

**Reply by ThePoipoi** [reply]
Hi

My config is almost the same as previously

```
    config:
      [...]
      log.retention.ms: 120000
      log.retention.bytes: 10737
      log.roll.ms: 120000
      log.delete.retention.ms: 120000
      log.cleaner.delete.retention.ms: 120000
      log.retention.check.interval.ms: 10000
```

And if I take a look at `/var/lib/kafka`, I can see a folder containing a kafka_log0 file:

```
[kafka@my-cluster-kafka-0 kafka]$ pwd
/var/lib/kafka
[kafka@my-cluster-kafka-0 kafka]$ ls
data
[kafka@my-cluster-kafka-0 kafka]$ cd data/
[kafka@my-cluster-kafka-0 data]$ ll
total 4
drwxr-xr-x. 57 kafka root 4096 Feb  1 10:41 kafka-log0
[kafka@my-cluster-kafka-0 data]$
```

And if I describe my (fake) consumer group to watch information about it (I have 5 partitions):

[tester@ktm376 ~]$ kubectl -n kafka-a exec -ti my-cluster-kafka-0 -- ./bin/kafka-consumer-groups.sh --bootstrap-server my-cluster-kafka-tls-bootstrap:9093 --describe --group console-consumer-8641

```
GROUP                 TOPIC           PARTITION  CURRENT-OFFSET  LOG-END-OFFSET  LAG             CONSUMER-ID                                                           HOST            CLIENT-ID
console-consumer-8641 syncTopicSDE    0          -               2120            -               consumer-console-consumer-8641-1-34c4d800-abaf-4095-b77b-42726afb1a66 /X.X.X.62    consumer-console-consumer-8641-1
console-consumer-8641 syncTopicSDE    4          -               1980            -               consumer-console-consumer-8641-1-34c4d800-abaf-4095-b77b-42726afb1a66 /X.X.X.62    consumer-console-consumer-8641-1
console-consumer-8641 syncTopicSDE    1          -               1980            -               consumer-console-consumer-8641-1-34c4d800-abaf-4095-b77b-42726afb1a66 /X.X.X.62    consumer-console-consumer-8641-1
console-consumer-8641 syncTopicSDE    3          -               1800            -               consumer-console-consumer-8641-1-34c4d800-abaf-4095-b77b-42726afb1a66 /X.X.X.62    consumer-console-consumer-8641-1
console-consumer-8641 syncTopicSDE    2          -               2120            -               consumer-console-consumer-8641-1-34c4d800-abaf-4095-b77b-42726afb1a66 /X.X.X.62    consumer-console-consumer-8641-1
```

**Reply by scholzj** [reply]
> And if I take a look at /var/lib/kafka, I can see a folder containing a kafka_log0 file:

You have to dive into the directory tree to find the directory for your topic/partition and check there if the files are created every 2 minutes or not.

**Reply by ThePoipoi** [reply]
I'm afraid that they didn't change since this morning, at their creation

```
drwxr-xr-x. 2 kafka root  144 Feb  1 10:42 syncTopicSDE-0
drwxr-xr-x. 2 kafka root  144 Feb  1 10:42 syncTopicSDE-1
drwxr-xr-x. 2 kafka root  144 Feb  1 10:42 syncTopicSDE-2
drwxr-xr-x. 2 kafka root  144 Feb  1 10:42 syncTopicSDE-3
drwxr-xr-x. 2 kafka root  144 Feb  1 10:42 syncTopicSDE-4
```

**Reply by scholzj** [reply]
These are the partition directories ... you have to check the files inside them.

**Reply by ThePoipoi** [reply]
Sorry, I was not clear enough, but it's the same inside each folder

```
[kafka@my-cluster-kafka-0 kafka-log0]$ cd syncTopicSDE-0
[kafka@my-cluster-kafka-0 syncTopicSDE-0]$ ls
total 8
-rw-r--r--. 1 kafka root        0 Feb  1 10:41 00000000000000002120.log
-rw-r--r--. 1 kafka root       10 Feb  1 10:41 00000000000000002120.snapshot
-rw-r--r--. 1 kafka root 10485756 Feb  1 10:41 00000000000000002120.timeindex
-rw-r--r--. 1 kafka root       11 Feb  1 10:41 leader-epoch-checkpoint
```

And it's confirmed by this command:
`kubectl -n kafka-a exec -ti my-cluster-kafka-0 -- ./bin/kafka-console-consumer.sh --bootstrap-server my-cluster-kafka-tls-bootstrap:9093 --topic syncTopicSDE --from-beginning`

Because it still list each message since this morning 10:41

But another thing: when I push more messages to this specific topic (syncTopicSDE), a new file is created in partition directories, named as the ".log" file but tagged as ".index".
This specific file is alive for 2/3 min then disapears.

**Reply by scholzj** [reply]
I guess the main question is whether these change every 2 minutes as you have it in configs now. Is it always only this 2120 file? Or are new files being created and deleted? 

One thing to keep in mind is that by default, the clients can set different timestamps. And if one of your clients sets a timestamp _in the future_, it might block the deletion. That can be controlled by `log.message.timestamp.difference.max.ms` and `log.message.timestamp.type`. Not sure it is your case or not of course.

I'm not sure I understand what does it mean _new file is created in partition directories, named as the ".log" file but tagged as ".index"_. But if it is deleted after 3-2 minutes that almost sounds like what you wanted? The new segements to be created and deleted often?

**Reply by scholzj** [reply]
Glad it works now! 👍

**Reply by ThePoipoi** [reply]
OK !!!

So finally, it seems to work. I wasn't understanding how files were managed, and I misunderstood information...
But it works ! If I run the previous command with "from-beginning", finally they disappeared. Not exactly after 2 min like you said, but a little bit more, waiting for the next loop to work (and don't worry, it's not a problem for us). I also think I've been confused by the timestamp value.

In fact, I have observed how the folder is managed right after sending new messages to Kafka:
- a new file 00000000000000004158.index is created
- after a few time, 3 more files are created (00000000000000004158.index.deleted 00000000000000004158.log.deleted and 00000000000000004158.timeindex.deleted). I assume that they have been tag as "can be deleted"
- after another time, all of those files are removed, and nothing is remaining in the topic (if I run an new consumer, it just waiting); just "LOG-END-OFFSET" are still remaining as previously with a bigger index


So so so sorry for losing your time, it was all my bad. Since I made the modification with `log.roll.ms`, everything is fine. I just mixed some information and made mistakes...
